### DISCUSSION TITLE
Values stored to ApplicationContext.ClientContext are not available on the server.

### DISCUSSION THREAD

**Discussion by adrianwright109** [MarimerLLC/csla#2213] [Questions]
I'm trying to store UI state in a Blazor Web Assembly component by using `Csla.ApplicationContext.ClientContext` as I've read it passes data from Client to Server.

The hope is that the UI state stored in `Csla.ApplicationContext.ClientContext` will be available on the DataPortal API when doing calls to `[Create]` `[Fetch]` etc methods without the need of passing the UI state values as parameters. 

I'm not sure what I've outlined above is possible as I've tried to do it but without any success.

Details of what I've tried so far...

I have a class called `AppState` which is in the `Shared` project so that it is accessible between `Client` and `Server` projects.

That class has been added to the `Client` DI in `Program.cs` so its available throughout the Blazor WebAssembly app.
```
builder.Services.AddScoped(typeof(AppState));
```

I have a `CascadingAppStateProvider` component which wraps the application so that state changes are handled in this component.

The `AppState` is being populated correctly in the component and then is being assigned to the `Csla.ApplicationContext.ClientContext`
```
@code {
    private const string AppStateSessionStorageKey = "AppState";

    public async Task SaveChangesAsync()
    {
        ApplicationContext.ClientContext[AppStateSessionStorageKey] = AppState;
    }
}
```
![image](https://user-images.githubusercontent.com/47532258/115051742-0ee45600-9ed5-11eb-9d33-bab4db14f748.png)

When trying to retrieve the same UI state on the server from a DataPortal API method `Csla.ApplicationContext.ClientContext` has an instance of `AppState` but it has all the `default` values and none of the values set in the Blazor component.
![image (1)](https://user-images.githubusercontent.com/47532258/115052939-7222b800-9ed6-11eb-957e-1b12df5ab3c4.png)

**Version and Platform**
CSLA version: 5.4.2
OS: Windows
Platform: ASP.NET Core, Blazor WebAssembly Hosted

`AppState.cs`
```
namespace MB5.Shared
{
    [Serializable]
    public class AppState : MobileObject
    {
        public int? ClientId { get; set; } = 0;     //TODO: Get from User

        public ClientEnvironment? ClientEnvironment { get; set; }

        public int? SessionId { get; set; }

        public int? SettingsGroupId { get; set; } = -1;

        public string SettingsGroupName { get; set; } = "Default";

        public string Identifier { get; set; }

        public void SetIdentifier(ComponentBase source, Guid guid)
        {
            Identifier = guid.ToString();
            NotifyStateChanged(source, nameof(Identifier));
        }

        public void SetClientId(ComponentBase source, int? clientId)
        {
            ClientId = clientId;
            NotifyStateChanged(source, nameof(ClientId));
        }

        public void SetClientEnvironment(ComponentBase source, ClientEnvironment? clientEnvironment)
        {
            ClientEnvironment = clientEnvironment;
            NotifyStateChanged(source, nameof(ClientEnvironment));
        }

        public void SetSessionId(ComponentBase source, int? sessionId)
        {
            SessionId = sessionId;
            NotifyStateChanged(source, nameof(SessionId));
        }

        public void SetSettingsGroup(ComponentBase source, int? settingsGroupId, string settingsGroupName)
        {
            SettingsGroupId = settingsGroupId;
            if (!string.IsNullOrWhiteSpace(settingsGroupName))
                SettingsGroupName = settingsGroupName;
            NotifyStateChanged(source, nameof(SetSettingsGroup));
        }

        public event Action<ComponentBase, string> StateChanged;

        private void NotifyStateChanged(ComponentBase source, string property) => StateChanged?.Invoke(source, property);

    }
}
```

**Comment by adrianwright109** [comment]
**Update**
It works if I just assign a string value to `Csla.ApplicationContext.ClientContext` that value persists to the `Server`. So looks to be my `AppState` class that it can't handle.

**Comment by rockfordlhotka** [comment]
The `ClientContext` is serialized using the formatter configured for the data portal, which is usually `MobileFormatter`. So any primitive type, or type that implements `IMobileObject` should work. Other types will not work.

**Comment by adrianwright109** [comment]
I have created a new app state class called `AppStateTest` to experiment with.  

`AppStateTest` class in full
```
    public interface IAppStateTest
    {
        string FirstName { get; set; }
        string Surname { get; set; }
    }

    [Serializable]
    public class AppStateTest : MobileObject, IAppStateTest
    {
        private string _firstName = "Fred";
        private string _surname = "Bloggs";

        public string FirstName
        {
            get { return _firstName; }
            set { _firstName = value; }
        }

        public string Surname
        {
            get { return _surname; }
            set { _surname = value; }
        }
    }
```

This `AppStateTest` class inherits from the `MobileObject` abstract class.  

I have also created my `public` properties with backing fields (I have tried both `public` and `private` backing fields) as described in https://github.com/MarimerLLC/csla/blob/main/docs/serialization.md#mobileformatter-constraints

To save the object to the `ClientContext` from a razor component I call:
```
Csla.ApplicationContext.ClientContext[nameof(AppStateTest)] = new AppStateTest { FirstName = "Adrian" };
```

To retrieve that same object I call:
```
var appStateTestObject = ApplicationContext.ClientContext.GetValueOrNull(nameof(AppStateTest));
```

When I retrieve the saved object when still inside the razor component I get the following returned:
![image](https://user-images.githubusercontent.com/47532258/115110516-bf089c00-9f73-11eb-9762-299f49ad9f09.png)

This is expected as I have overwritten the `FirstName`'s default value of `Fred` with `Adrian`.

If I do the same retrieve call on the `Server` in a `[Fetch]` method for example I'm expecting to get the same values as I did in the razor component.

Actual result is I don't get the same values instead I just seem to get the `AppStateTest` object in its default state with its default values as shown below:
![image](https://user-images.githubusercontent.com/47532258/115110648-5c63d000-9f74-11eb-8c7d-d2bf52f9ac7f.png)

Am I missing something here ?

**Reply by rockfordlhotka** [reply]
You need to use managed backing fields, or override the get/set state methods and manually add/retrieve the backing field values from the serializationinfo object.

**Comment by rockfordlhotka** [comment]
Also, this thread may have useful info about an issue that's fixed in v5.4.2.

https://github.com/MarimerLLC/csla/discussions/2153
